amr-wind/physics/TerrainDrag.cpp

```diff
@@ -0,0 +1,180 @@
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_xrough.begin(), m_xrough.end(),
+            gpu_xrough.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_yrough.begin(), m_yrough.end(),
+            gpu_yrough.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_z0rough.begin(), m_z0rough.end(),
+            gpu_z0rough.begin());
+        const auto* xrough_ptr = gpu_xrough.data();
+        const auto* yrough_ptr = gpu_yrough.data();
+        const auto* z0rough_ptr = gpu_z0rough.data();
+        for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+            const auto& vbx = mfi.validbox();
+            auto levelBlanking = blanking.array(mfi);
+            auto levelDrag = drag.array(mfi);
+            auto levelz0 = terrainz0.array(mfi);
+            const unsigned terrainSize = m_xterrain.size();
+            const unsigned roughnessSize = m_xrough.size();
+            amrex::ParallelFor(
+                vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    // compute the source term
+                    const amrex::Real x1 = prob_lo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real x2 = prob_lo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real x3 = prob_lo[2] + (k + 0.5) * dx[2];
+                    // Terrain Height
+                    amrex::Real residual = 10000;
+                    amrex::Real terrainHt = 0.0;
+                    for (unsigned ii = 0; ii < terrainSize; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x1 - xterrain_ptr[ii], 2) +
+                            std::pow(x2 - yterrain_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            terrainHt = zterrain_ptr[ii];
+                        }
+                        if (radius < dx[0]) {
+                            break;
+                        }
+                    }
+                    int turnOn = (x3 <= terrainHt) ? 1 : 0;
+                    levelBlanking(i, j, k, 0) = turnOn;
+                    residual = 10000;
+                    amrex::Real roughz0 = 0.1;
+                    for (unsigned ii = 0; ii < roughnessSize; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x1 - xrough_ptr[ii], 2) +
+                            std::pow(x2 - yrough_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            roughz0 = z0rough_ptr[ii];
+                        }
+                        if (radius < dx[0]) {
+                            break;
+                        }
+                    }
+                    levelz0(i, j, k, 0) = roughz0;
+                });
+            amrex::ParallelFor(
```

Comment:
Is there a reason for two separate `ParallelFor` loops here? Both this one and the one above look to be on the same vbx

Reply:
The terrainDrag ParallelFor call depends on the terrainBlank variable to be setup. This was the reason that I needed two calls.